### PR TITLE
Add test statistics

### PR DESCRIPTION
**Pull request by matthewreyna**
These changes add the ability to use and define different test statistics with WExT. The choice of test statistic is exposed as command-line argument for the `find_sets.py` script and defined in the check_condition function of the `wext/saddlepoint.py` script. Currently, three test statistics are available, but others are easy to define:

1. Exclusivity (`exclusivity`): the number of samples with mutually exclusive mutations (exactly one mutated gene in a gene set M).
2. Any co-occurrence (`any-co-occurrence`): the number of samples with any co-occurring mutations (at least two mutated genes in a gene set M).
3. All co-occurrence (`all-co-occurrence`): the number of samples with all co-occurring mutations (exactly k mutated genes in a gene set M of size k).

There is a simple example in the `examples/simple` directory that uses these test statistics, and it runs in a couple of seconds. Also, the `find_sets.py` script only supports the WRE test with the enumeration search strategy, but the existing `find_exclusive_sets.py` script is unchanged.

A new [entry on the wiki](https://github.com/raphael-group/wext/wiki/Find-sets) describes these options and explicitly defines each of these statistics.

### CODE REVIEW

File: wext/saddlepoint.py [hunk 1]
```diff
@@ -5,16 +5,29 @@
...
 from constants import *
 
-def condition(state):
+def check_condition(state, condition):
 
     # Define a success condition for each state, e.g., mutual exclusivity.
 
-    if sum(state)==1:
-        return True
+    if condition=='exclusivity':
```
Comment (mdml) [inline wext/saddlepoint.py]: Would prefer if these conditions were constants (e.g. `EXCLUSIVITY='exclusivity'` and `if condition == EXCLUSIVITY`).

File: wext/exclusivity_tests.py [hunk 2]
```diff
@@ -59,3 +59,25 @@ def re_test(t, x, tbl, method=EXACT, verbose=0):
         p_value, mid_p_value = comet_exact_test( k, N, tbl, 1.1 )
 
     return p_value
+
+def general_wre_test(gene_set, geneToCases, p, condition, verbose=0):
```
Comment (mdml) [inline wext/exclusivity_tests.py:63]: Could use some documentation

File: examples/simple/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+a, b	0.000227547596931	0.0113258255022	0.00342392921448	12	0	2	6	6	0
+```
+
+The co-occurrence results (`any-co-occurrence_results-k2.tsv` and `all-co-occurrence_results-k2.tsv`) should be very similar to the following table, which shows that only genes `c` and `c` have co-occurring mutations with the default FDR < 0.5.
+```
+#Gene set	WRE (Saddlepoint) P-value	WRE (Saddlepoint) FDR	WRE (Saddlepoint) Runtime	T	Z	t00	t01	t10	t11
+c, d	0.00351621696959	0.175014196425	0.00170493125916	0	9	5	0	0	9
+```
+
+The WExT wiki describes the details of the above tests and the arguments for the above scripts in more detail.
```
Comment (mdml) [inline examples/simple/README.md]: Maybe add a link to WExT wiki?

File: examples/simple/README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+10	b
+11	a	e
+12	b	g
+13	c   d
+14  c   d
+```
+
+## Output
+
+The mutual exclusivity results (`excusivity_results-k2.tsv`) should be very similar to the following table, which shows that only genes `a` and `b` have mutually exclusive mutations with the default FDR < 0.5.
```
Comment (mdml) [inline examples/simple/README.md]: "excusivity" --> "exclusivity"

File: examples/simple/commands.sh [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+num_cores=4
+
+# Preprocess mutations.
+python ../../process_mutations.py \
+    -m  adjacency_list.tsv \
+    -ct NA \
+    -o  data.json
+
+# Compute mutation probabilities.
+python ../../compute_mutation_probabilities.py \
```
Comment (mdml) [inline examples/simple/commands.sh:13]: Can we provide a random seed so the results are reproducible?

**Review by mdml — CHANGES_REQUESTED** [review]
I ran the examples and get mostly similar results, and overall the code looks good! I have a few minor nits that I commented about separately.

The only thing I don't quite understand is the large amount of code duplication -- both for `find_exclusive_sets.py` and `find_sets.py`, and `test_set_group` and `general_test_set_group`. This certainly isn't the ideal solution, so I assume it is a somewhat temporary fix. Perhaps we can add an issue to the fact effect?

File: examples/simple/README.md (hunk 4 above)
Comment (matthewreyna) [reply examples/simple/README.md]: Done.

File: examples/simple/README.md (hunk 3 above)
Comment (matthewreyna) [reply examples/simple/README.md]: Done, including separate links for each script.

File: examples/simple/commands.sh (hunk 5 above)
Comment (matthewreyna) [reply examples/simple/commands.sh:13]: We already provide a random seed (the index for the each permutation), so the results will be reproducible on the same machine, but they will not necessarily be consistent from machine to machine.

File: wext/exclusivity_tests.py (hunk 2 above)
Comment (matthewreyna) [reply wext/exclusivity_tests.py:63]: I updated the wiki.

File: wext/saddlepoint.py (hunk 1 above)
Comment (matthewreyna) [reply wext/saddlepoint.py]: Done.  Note that constants add an additional step when users to define their own test statistics.

**Comment by matthewreyna** [comment]
I updated the code to address all of the above nits.

The code duplication is a temporary byproduct of how exclusivity is somewhat tightly integrated in some of the code base.  This issue should resolve itself more easily in the future as we update or deprecate some features, e.g., the recursive formula, but I wanted to provide these extra features more expediently.

File: examples/simple/commands.sh (hunk 5 above)
Comment (mdml) [reply examples/simple/commands.sh:13]: Oh, I see. That still seems weird. So the seeds always start at 0 (or 1)?

Comment (matthewreyna) [reply examples/simple/commands.sh:13]: Yes, the random seeds for the `n` permuted matrices are `i`, `i+1`, ..., `i+n-1`, where `i` is user-definable (the argument `start_index`) but 1 by default.  We can add an optional `seed` argument and use that to generate seeds for each of the permuted matrices, which is less opaque but similar enough in practice.  However, we still cannot guarantee that different machines with different versions of Python, different gcc compilers, etc. will generate the same random numbers from the same random seed.

Comment (mdml) [reply examples/simple/commands.sh:13]: I understand making the results the exact same on different machines can be challenging. My point was more that we should make the random seed more (obviously) accessible.

Comment (matthewreyna) [reply examples/simple/commands.sh:13]: Done.  I added an optional argument (`-s`, `--seed`).  By default, it picks a random seed, but users can choose their own seed.  I updated the example to use a random seed, and I updated the documentation to describe it.

**Comment by mdml** [comment]
All looks good. I vote merge when ready.

**Pull request merged**